tests-integration/poller_test.go

```diff
@@ -461,3 +463,75 @@ func TestPollerExpiryEnsurePollingRace(t *testing.T) {
 		t.Fatalf("Should have got 401 http response; got %d\n%s", status, resBytes)
 	}
 }
+
+// Regression test for the bugfix for https://github.com/matrix-org/sliding-sync/issues/287#issuecomment-1706522718
+// Specifically, we could cache the failure and never tell the poller about new tokens, wedging the client(!). This
+// seems to have been due to the following:
+//   - client hits sync for the first time. We /whoami and remember the token->user mapping in TokensTable.
+//   - client syncing + poller syncing, everything happy.
+//   - token expires. OnExpiredToken is sent to EnsurePoller which removes the entry from EnsurePoller and nukes the conns.
+//   - client hits sync, gets 400 M_UNKNOWN_POS due to nuked conns.
+//   - client hits a fresh /sync: for whatever reason, the token is NOT 401d there and then by the /whoami lookup failing.
+//     Maybe failed to remove the token, but don't see any logs to suggest this. Seems to be an OIDC thing.
```

Comment:
sad that we don't have a better understanding of this. Is there any logging we can add to make things less unclear for our future selves?

Reply:
nothing comes to mind immediately sadly.